linkerd/app/outbound/src/ingress.rs

```diff
@@ -79,142 +88,143 @@ impl<H> Outbound<H> {
+                            metadata: Metadata::default(),
+                            logical_addr: None,
+                            protocol: version,
+                            opaque_protocol: false,
+                            tls: tls::ConditionalClientTls::None(
+                                tls::NoClientTls::IngressWithoutOverride,
+                            ),
+                        }));
+                    }
+
+                    // Otherwise, if a profile was discovered, use it to build a logical stack.
+                    if let Some(profile) = profile {
+                        let addr = profile.borrow().addr.clone();
+                        if let Some(logical_addr) = addr {
+                            return Ok(svc::Either::A(http::Logical {
+                                profile,
+                                logical_addr,
+                                protocol: version,
+                            }));
+                        }
+                    }
+
+                    // Otherwise, the override header was present but no profile information could
+                    // be discovered, so fail the request.
+                    Err(ProfileRequired)
+                },
+                http_endpoint,
+            )
+            .push(profiles::discover::layer(profiles, move |h: Http| {
+                // Lookup the profile if the override header was set and it is in the configured
+                // profile domains. Otherwise, profile discovery is skipped.
+                if let Target::Override(dst) = h.target {
+                    if profile_domains.matches(dst.name()) {
+                        return Ok(profiles::LookupAddr(dst.into()));
+                    }
+                }
 
-    fn recognize(&self, req: &http::Request<B>) -> Result<Self::Key, Error> {
-        let dst = http_request_l5d_override_dst_name_addr(req).map_err(|_| DstOverrideRequired)?;
-        Ok(Target {
-            dst,
-            version: self.0.protocol,
-        })
+                tracing::debug!(
+                    domains = %profile_domains,
+                    "Address not in a configured domain",
+                );
+                Err(profiles::DiscoveryRejected::new(
+                    "not in configured ingress search addresses",
+                ))
+            }))
+            // This service is buffered because it needs to initialize the profile resolution and a
+            // fail-fast is instrumented in case it becomes unavailable. When this service is in
+            // fail-fast, ensure that we drive the inner service to readiness even if new requests
+            // aren't received.
+            .push_on_response(
+                svc::layers()
+                    .push(rt.metrics.stack.layer(stack_labels("http", "logical")))
+                    .push(svc::layer::mk(svc::SpawnReady::new))
+                    .push(svc::FailFast::layer("HTTP Logical", dispatch_timeout))
```

Comment:
nit: if i'm reading this correctly it looks like we will produce stack metrics for "HTTP logical" even on forwards --- shouldn't these stack metrics be on the logical stack inside of the switch, instead?